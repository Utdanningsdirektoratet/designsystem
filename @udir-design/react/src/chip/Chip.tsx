export {
  Chip,
  ChipButton,
  type ChipButtonProps,
  ChipCheckbox,
  type ChipCheckboxProps,
  ChipRadio,
  type ChipRadioProps,
  ChipRemovable,
  type ChipRemovableProps,
} from '@digdir/designsystemet-react';
