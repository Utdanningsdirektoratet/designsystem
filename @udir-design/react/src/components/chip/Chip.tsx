import {
  Chip,
  ChipButton,
  type ChipButtonProps,
  ChipCheckbox,
  type ChipCheckboxProps,
  ChipRadio,
  type ChipRadioProps,
  ChipRemovable,
  type ChipRemovableProps,
} from '@digdir/designsystemet-react';

// For some reason this fixes "ComponentSubcomponent" -> "Component.Subcomponent" in Storybook code snippets
Chip.Button.displayName = 'Chip.Button';

export type {
  ChipButtonProps,
  ChipCheckboxProps,
  ChipRadioProps,
  ChipRemovableProps,
};
export { Chip, ChipButton, ChipCheckbox, ChipRadio, ChipRemovable };
