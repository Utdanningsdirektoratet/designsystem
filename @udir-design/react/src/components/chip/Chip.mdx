import { Meta, Canvas, Controls, Primary } from '@storybook/addon-docs/blocks';
import { ChipEx1 } from './docs/doAndDont';
import * as ChipStories from './Chip.stories';

<Meta of={ChipStories} />

# Chip

`Chip` er ein liten, interaktiv komponent som lèt brukarane styre korleis dei vil sjå innhald. For eksempel kan den brukast til å filtrere kategoriar i eit søkeresultat og vise kva for filter som er aktive.

**Passar til å**

- filtrere informasjon, for eksempel i lister eller tabellar.
- vise kva for val som er aktive.

**Passar ikkje til å**

- indikere varsel for tal eller status (bruk heller [`Badge`](/docs/components-badge--docs)).
- merke innhald med kategori/emne som ikkje er interaktivt. Bruk heller [`Tag`](/docs/components-tag--docs)-komponenten dersom kategori/emne ikkje er klikkbart.
- brukast i menyar.

## Slik brukar du Chip

<Primary />
<Controls />

```tsx
import { Chip } from '@udir-design/react';

// som radio
<Chip.Radio name="my-chip" defaultChecked>
  Bokmål
</Chip.Radio>;

// som checkbox
<Chip.Checkbox name="my-chip" defaultChecked>
  Barnehage
</Chip.Checkbox>;

// som removable
<Chip.Removable>Grunnskole</Chip.Removable>;

// som button
<Chip.Button>Vidregående</Chip.Button>;
```

## Varianter

### Chip.Button

`Chip.Button` kan brukast i gruppe for å vise valalternativ.

<Canvas of={ChipStories.Button} />

### Chip.Checkbox

`Chip.Checkbox` kan brukast som eit alternativ til `Checkbox`.

<Canvas of={ChipStories.Checkbox} />

### Chip.Radio

`Chip.Radio` kan brukast som eit alternativ til `Radio` når brukaren kun kan bruke eit filter av ganga.

<Canvas of={ChipStories.Radio} />

### Chip.Removable

`Chip.Removable` inneheld eit kryss som indikerer at filteret kan fjernast. Det er viktig å merke seg at `aria-label` må leggjast til dersom innhaldet i Chip ikkje forklarer at den kan fjernast. I dette eksempelet er det lagt til `aria-label="Slett {nivå}"`.

<Canvas of={ChipStories.Removable} />

## Retningslinjer

Unngå å plassere `Chips` vertikalt. Skal du vise alternativ under kvarandre, bruk heller [`Checkbox`](/docs/components-checkbox--docs) eller [`Radio`](/docs/components-radio--docs). En gruppe med flere `Chips` kan brytast over fleire linjer.

<ChipEx1 />

## Tekst i Chip

`Chip` bør ha så få ord som mogleg, helst berre eitt eller to.

## Tilgjengelegheit

Viss delar av ei side blir oppdaterte samtidig som brukarane gjer noko på sida, må vi passe på at brukarar med skjermlesarar får beskjed om det. Vi kan bruke ein `ARIA live region` til det, som kan vise tal på treff eller andre endringar.
