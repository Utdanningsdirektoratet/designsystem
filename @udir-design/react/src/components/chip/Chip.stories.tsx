import type { Meta, StoryObj } from '@storybook/react-vite';
import { Chip } from './Chip';
import { expect, fn, within } from '@storybook/test';

const meta: Meta<typeof Chip.Radio> = {
  component: Chip.Radio,
  tags: ['alpha'],
  parameters: {
    customStyles: { display: 'flex', gap: 'var(--ds-size-2)' },
  },
};

export default meta;
type Story = StoryObj<typeof Chip.Radio>;

export const Preview: Story = {
  render: (args) => (
    <>
      <Chip.Radio {...args} name="my-radio" value="nynorsk" defaultChecked>
        Nynorsk
      </Chip.Radio>
      <Chip.Radio {...args} name="my-radio" value="bokmål">
        Bokmål
      </Chip.Radio>
    </>
  ),
  play: async ({ canvasElement, step }) => {
    const canvas = within(canvasElement);
    const inputNynorsk = canvas.getByRole('radio', { name: 'Nynorsk' });
    const inputBokmal = canvas.getByRole('radio', { name: 'Bokmål' });

    await step('Render label and input as native element', async () => {
      const label = canvas.getByLabelText('Nynorsk');
      expect(label).toBeInTheDocument();
      expect(inputNynorsk).toBeInTheDocument();
    });

    await step('Select Nynorsk', async () => {
      inputNynorsk.click();
      expect(inputNynorsk).toBeChecked();
      expect(inputBokmal).not.toBeChecked();
    });

    await step('Select Bokmal', async () => {
      inputBokmal.click();
      expect(inputNynorsk).not.toBeChecked();
      expect(inputBokmal).toBeChecked();
    });
  },
};

export const Checkbox: StoryObj<typeof Chip.Checkbox> = {
  render: (args) => <Chip.Checkbox {...args}>Nynorsk</Chip.Checkbox>,
  play: async ({ canvasElement, step }) => {
    const canvas = within(canvasElement);
    const input = canvas.getByRole('checkbox');

    await step('Render label and input as native element', async () => {
      const label = canvas.getByLabelText('Nynorsk');
      expect(label).toBeInTheDocument();
      expect(input).toBeInTheDocument();
    });

    await step('Toggle checkbox', async () => {
      input.click();
      expect(input).toBeChecked();
      input.click();
      expect(input).not.toBeChecked();
    });
  },
};

export const Removable: StoryObj<typeof Chip.Removable> = {
  args: {
    'aria-label': 'Slett Norge',
    onClick: fn(),
  },
  render: (args) => <Chip.Removable {...args}>Norge</Chip.Removable>,
  play: async ({ canvasElement, step, args }) => {
    const canvas = within(canvasElement);
    const button = canvas.getByRole('button');

    await step('Render label and button as native element', async () => {
      const label = canvas.getByText('Norge');
      expect(label).toBeInTheDocument();
      expect(button).toBeInTheDocument();
    });

    await step('Remove chip', async () => {
      button.click();
      expect(args.onClick).toHaveBeenCalled();
    });
  },
};

export const Button: StoryObj<typeof Chip.Button> = {
  args: {
    onClick: fn(),
  },
  render: (args) => (
    <>
      <Chip.Button {...args}>Søk etter nynorsk</Chip.Button>
      <Chip.Button {...args}>Søk etter bokmål</Chip.Button>
      <Chip.Button {...args}>Søk etter engelsk</Chip.Button>
    </>
  ),
  play: async ({ canvasElement, step, args }) => {
    const canvas = within(canvasElement);
    const buttons = canvas.getAllByRole('button');

    await step('Render label and button as native element', async () => {
      const label = canvas.getByText('Søk etter nynorsk');
      expect(label).toBeInTheDocument();
      expect(buttons).toHaveLength(3);
    });

    await step('Click button', async () => {
      buttons[1].click();
      expect(args.onClick).toHaveBeenCalled();
    });
  },
};
