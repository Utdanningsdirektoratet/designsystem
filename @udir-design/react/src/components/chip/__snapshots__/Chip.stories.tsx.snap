// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button 1`] = `
<button type="button">
  Søk etter nynorsk
</button>
<button type="button">
  Søk etter bokmål
</button>
<button type="button">
  Søk etter engelsk
</button>
`;

exports[`Checkbox 1`] = `
<label>
  <input type="checkbox">
  Nynorsk
</label>
`;

exports[`Preview 1`] = `
<label>
  <input
    type="radio"
    name="my-radio"
    value="nynorsk"
    checked
  >
  Nynorsk
</label>
<label>
  <input
    type="radio"
    name="my-radio"
    value="bokmål"
  >
  Bokmål
</label>
`;

exports[`Removable 1`] = `
<button
  type="button"
  data-removable="true"
  aria-label="Slett Norge"
>
  Norge
</button>
`;
